Feature/timer stop resume

Minny27 opened the pull request:
## Motivation ⍰

- 메인 화면 타이머 스탑, 재생 구현

<br>

## Key Changes 🔑

- 더블 클릭 (일시 정지), 다른 화면 이동까지 구현
- 클릭 이벤트, viewWillAppear, viewWillDisAppear 합성으로 true false 리턴하는 방식으로 구현
- rx 확장 코드 삭제
- vc에서 cell을 생성할 때 cell viewmodel에 timerActiveTrigger를 input으로 전달하고
- cell의 indexpath.row와 scrollindex가 같을 때만 bind
- timerActiveTrigger가 포함하고 있는 bool값이 false(멈춤)이면 최근 emit한 값만 리턴
- timerActiveTrigger가 포함하고 있는 bool값이 true(재개)이면 최근 emit한 값을 기준으로 타이머 재생

주요 로직
```swift
  func transform(input: Input) -> Output {
    var currentTime: Double = 8.0 // 타이머에서 emit되는 최근 값 저장
    var startTime: Double = 8.0 // 고정 시작 시간
    let user = Driver.just(self.userDomain)
    
    let timerActiveTrigger = input.timerActiveTrigger
      .startWith(true) // timerActiveTrigger를 전달받아서 사용하다보니 기본 값은 true로 바로 실행
      .asObservable()
    
    let timer = timerActiveTrigger
      .flatMapLatest { value in
        if !value {
          if currentTime == 0 { currentTime = 8.0 } // 다음 셀로 넘어가도 0이면 자동 스크롤되므로 0초가 되면 다시 8.0으로 갱신
          startTime = currentTime 
          return Driver.just(currentTime)
        } else {
          return Observable<Int>.interval(.milliseconds(10),
                                          scheduler: MainScheduler.instance)
          .take(Int(startTime * 100) + 1) // 시간의 총 개수
          .map { value in
            let time = round((startTime * 100) - Double(value)) / 100
            currentTime = time
            return time
          }
          .asDriver(onErrorDriveWith: Driver<Double>.empty())
        }
      }.asDriver(onErrorJustReturn: 8.0)
{ ... }
}
```

미리보기
<img src="https://github.com/THT-Team/THT-iOS/assets/68800789/9fce0ffd-59e9-487a-8856-a161e830acd6" width=25%>

<br>

## To Reviewers 🙏🏻

- [x] 더블 클릭 시 정지, 재개 잘 되는 지 봐주세요
- [x] 다른 화면으로 이동 후 다시 돌아왔을 때, 시간이 유지되는지 봐주세요
- cell init할 때 현재는 bind, bindViewModel 두함수로 할당하고
- 셀에서 Scrollindex와 동일할 때만 bindViewModel을 실행하도록 구독하고 있는데
- 셀 안에서 구독을 하지 않도록 하려고 시도를 많이했는데 잘 안 되서 한 번 봐주시면 감사하겠습니다.

<br>

## Linked Issue 🔗

- [x] #36

Minny27 committed:
[#43] Refactor: DimViewManager, Card Button custom

- DimViewManager 구현으로 특정 뷰 위에 DimView를 추가할 수 있도록 구현
- DimView의 크기를 커스텀해서 특정 뷰에 올릴 수 있도록 커스텀함
- 세가지 Card Button이 같은 성격을 띄고 있어서 커스텀함

Minny27 committed:
[#36]Feat: 타이머 stop, resume 구현

- Rx+Utils에 pausable, pausableBuffered 확장
- vc에서 중단 트리거 머지해서 cell viewmodel에 전달

Minny27 committed:
[#36]Feat: 타이머 Start, Stop, Resume 구현

- 더블 클릭 (일시 정지), 다른 화면 이동까지 구현
- 클릭 이벤트, viewWillAppear, viewWillDisAppear 합성으로 true false 리턴하는 방식으로 구현
- rx 확장 코드 삭제
- vc에서 cell을 생성할 때 cell viewmodel에 timerActiveTrigger를 input으로 전달하고
- cell의 indexpath.row와 scrollindex가 같을 때만 bind
- timerActiveTrigger가 포함하고 있는 bool값이 false(멈춤)이면 최근 emit한 값만 리턴
- timerActiveTrigger가 포함하고 있는 bool값이 true(재개)이면 최근 emit한 값을 기준으로 타이머 재생

ibcylon commented on Falling/Sources/Feature/Main/Cell/MainCollectionViewCell.swift:
```diff
@@ -78,20 +48,23 @@ final class MainCollectionViewCell: TFBaseCollectionViewCell {
     profileCarouselView.tagCollectionView.isHidden = true
```
delegate 도 nil로 초기화하면 좋을 것 같습니다.

Minny27 replied:
반영하겠습니다!

ibcylon commented on Falling/Sources/Feature/Main/Cell/MainCollectionViewItemViewModel.swift:
```diff
@@ -116,15 +116,41 @@ final class MainCollectionViewItemViewModel: ViewModelType {
-      .asDriver(onErrorDriveWith: Driver<Double>.empty())
-
-    let timeState = time.map { TimeState(rawValue: $0) }
-    let isTimeOver = time.map { $0 == 0.0 }
```
해당 값이 true 일 때만 쓰이는데 필터걸어서 true 일 때만 방출되게 하면
 output 사용하는 쪽에서 if 쓸 필요가 없어질 것 같습니다.

Minny27 replied:
그러네요. 수정하겠습니다

ibcylon reviewed (approved):
to reviewer에서 cell 안에서 구독하고 있는 게 어떤게 있나요?

Minny27 committed:
[#44]Refactor: 필요없는 주입 및 코드 수정

Minny27 committed:
[#44]Refactor: DimView base view에 추가

- DimViewManager 삭제
- self를 참조해서 hide를 하다보니 뷰 안에서만 hide를 할 수 있어서
- base에 포함하게 됨.

Minny27 committed:
[#44]Enhancement: start, resume 동작 개선

- timmerActiveTrigger의 bool 값에
- viewWillAppear 호출 시, true, viewWillDisAppear 호출 시, false
- 이 둘을 머지해서 timerActiveRelay 값 갱신
- 카드 Double 탭은 timerActiveRelay의 값을 반대로 바꿔줌
- cell에 bind할 때, userCardScrollIndex와 timmerActiveTrigger 전달
- userCardScrollIndex가 갱신될 때, 최신 timmerActiveTrigger값을 합성하기위해
- withLatestFrom 사용
- 현재 셀의 index와 userCardScrollIndex가 같으면 dimview를 없애고
- timerActiveTrigger의 값을 리턴
- 같지 않으면 false를 리턴

Minny27 committed:
[#44]Refactor: 리뷰 내용 개선

- delegate = nil로 초기화
- isTimeOver를 timeZero로 이름 변경
- timeZero에서 timer를 filter로 0.0일 때만 emit

Minny27 commented:
현재까지 구현한 동작 미리보기입니다.
<img src="https://github.com/THT-Team/THT-iOS/assets/68800789/8d712506-de13-4dde-86e4-643eb68e0bae" width=25%>

ibcylon committed:
Refactor: weak 캡처 / do 제거

ibcylon committed:
Refactor: 불필요한 map 제거
